### ISSUE TITLE
Organizations: Assign Identity Provider to more than one organization.

### ISSUE THREAD

**Issue by osajdapawel** [keycloak/keycloak#31705]
### Description

Add possibility to link Identity Provider to multiple organization. 
At the moment Identity Provider can be linked only to one organization. It would be useful to link for example global Microsoft IdP to many organizations and filter the users based on company domain name.

### Motivation

At the moment if we want to use for instance Microsoft SSO within application and split users into organizations and allows user to authenticate with M365 account we need to create multiple the same IdProviders (one per organization). Then we assign IdProvider to organization and choose domain for org.

### Details

Would be interesting to discuss when we have possibility to link IdProvider to many organizations:

- show on login page
- redirect when domain matches (filter by domain: if we want more email domains team1.contoso.com, team2.contoso.com, IMO we could assing multiple times the same IdProvider to org and specify separate domains)

**Comment by ggomarighetti** [comment]
In a B2B SaaS application using the Organizations feature to separate customers, if we use GitHub or Google as login methods, we must configure a separate Identity Provider (IdP) for each customer.

This is not only relevant to internal applications or workforce-related use cases but is also critical for end users of the service. This aligns with the direction outlined in issue [#30180](https://github.com/keycloak/keycloak/issues/30180) to evolve Keycloak towards a "Customer Identity and Access Management (CIAM) with a focus on Business-to-Business (B2B) use cases."

**Comment by FredrikPAC** [comment]
Hi, great feature!
Just wanted to give my upvote.
We are a B2B SaaS looking at keycloak as a solution. We have umbrella organizations wanting to connect to our service, where the high level organization provide the idp while the sub level organizations are treated as tenants in our application based on email domain. Excited to follow the development on this.

**Comment by wschomburg** [comment]
Hi, we have the same use case where we have multiple organizations using the same IDP.

**Comment by bcmedeiros** [comment]
Is this currently planned?

**Comment by wschomburg** [comment]
I started to create dedicated IDPs for each organization and with Terraform it worked quite well. Just a few lines of config and a new organization with a dedicated IDP were configured in Keycloak. They all use the same Azure multi-tenant app configuration. Then I realized that Azure doesn't support wildcards in the redirect URLs and I have to add every IDP redirect URL in Azure, and it is also limited to 255 URLs per Azure application. 
So, having separate IDPs with the same configuration for each organization is also a limited solution. I'll go for only one organization and IDP for all customers using our multi-tenant Azure application for login for now.

Is there a plan to implement this feature in Keycloak?

**Comment by gscokart** [comment]
This feature is now in the top 15 of the requested features.  

I'm curious about the reasons why the restriction of one IDP to one organization only was initially introduced.  
Is there any authentication flow supported where we identify the user based on the idp identity first without having previously identified the organization ?  This would looks strange to me.

**Comment by gscokart** [comment]
> I started to create dedicated IDPs for each organization and with Terraform it worked quite well. Just a few lines of config and a new organization with a dedicated IDP were configured in Keycloak. They all use the same Azure multi-tenant app configuration. Then I realized that Azure doesn't support wildcards in the redirect URLs and I have to add every IDP redirect URL in Azure, and it is also limited to 255 URLs per Azure application. So, having separate IDPs with the same configuration for each organization is also a limited solution. I'll go for only one organization and IDP for all customers using our multi-tenant Azure application for login for now.
> 

@wschomburg , did you tried the workaround of adding all domains into a single Azure organization?  It would completely defeat the purpose and the benefits of organization attributes to handle multi tenancy and will force you to manage the tenant attribute at member level.  But I guess it would at least automatically redirect the user toward the right idp.

Still the feature the feature is required to have proper multi-tenancy support.

**Comment by gscokart** [comment]
One special situation that will need to be tested if same idp is used in 2 organization is a situation where :

2 keycloak users with 2 different email, from 2 different organization linked to the same idp are actually linked to the same federated_user_id.

I think the DB allows such situation, and I guess this situation is possible when the idp doesn't provide the email.

**Comment by wschomburg** [comment]
> [@wschomburg](https://github.com/wschomburg) , did you tried the workaround of adding all domains into a single Azure organization? It would completely defeat the purpose and the benefits of organization attributes to handle multi tenancy and will force you to manage the tenant attribute at member level. But I guess it would at least automatically redirect the user toward the right idp.
> 
> Still the feature the feature is required to have proper multi-tenancy support.

@gscokart Yes, that's what I'm doing now. One Azure organization with all customer email domains works for the automatic redirect, but all the other features of organizations couldn't be used (attribute or customer off-boarding by deleting the organization).

**Comment by gscokart** [comment]
Supporting this has an impact on the current model (and in DB and in REST API representation).  Currently, the idp has one optional attribute which is the organization.  The idp is thus really owned by the organization.  This model support having one organization owning multiple idp.  
This enhancement will requires using an n-n relationhip.

Note that this relationship will also have some attributes (new "kc.org.broker.redirect.mode.email-matches" and "kc.org.domain" are config element of the idp itself).

**Comment by wschomburg** [comment]
Sure, an n-to-m relationship is a more realistic use case for IDPs in the B2B context than the current 1-to-n relationship.

**Comment by gscokart** [comment]
Note that there is already some (maybe too limited) possibilities to use a single idp in multiple organisation.  The idp that are not bound to an organization but stay at realm level can still be used by any member of any organisation.
The organization can not restrict the potential list.  But this might already help in supporting some scenarios.

**Comment by vramik** [comment]
I've tried to capture requirements and behavioral decisions related to this issue as well as #47554 in https://github.com/keycloak/keycloak/discussions/49091. Please take a look and share your thoughts there.

**Comment by sulvund** [comment]
This is a super relevant feature for our use case as an SaaS provider and MSP for a large number of Entra ID customers. I just wanted to show our support and appreciation for the work being done here!
